File: Python/CountingValleys.py
# https://www.hackerrank.com/challenges/counting-valleys/problem?isFullScreen=true

class Solution:
    def countingValleys(self, steps: int, path: str) -> int:
        level = 0
        valleys = 0
        for step in path:
            if step == 'U':
                level += 1
                if level == 0:
                    valleys += 1
            else:
                level -= 1
        return valleys
    
if __name__ == "__main__":
    sol = Solution()
    path = "UDDDUDUU"
    steps = len(path)
    result = sol.countingValleys(steps, path)
    print(f"Number of valleys: {result}")

File: CountingValleys.java
// https://www.hackerrank.com/challenges/counting-valleys/problem?isFullScreen=true

class CountingValleys{
    public static int countingValleys(int steps, String path) {
        int level = 0;
        int valleys = 0;
        for (int i = 0; i < path.length(); i++){
            if(path.charAt(i) == 'U'){
                level++;
                if(level == 0){
                    valleys++;
                }
            }
            else if(path.charAt(i) == 'D'){
                level--;
            }
        }
        return valleys;
    }

    public static void main(String[] args) {
        String path = "UDDDUDUU";
        int steps = path.length();
        int result = countingValleys(steps, path);
        System.out.println("Number of valleys: " + result);
    }
}